### PR TITLE
source cleanup: remove all custom typedef structs

### PR DESCRIPTION
**Pull request by bagder**
- Stick to a single unified way to use structs
- Make checksrc complain on 'typedef struct {'
- Allow them in tests, public headers and examples

### CODE REVIEW

File: lib/curl_md5.h [hunk 1]
```diff
@@ -33,30 +33,30 @@ typedef void (* Curl_MD5_update_func)(void *context,
...
+struct MD5_context {
+  const struct MD5_params *md5_hash;    /* Hash function definition */
   void                  *md5_hashctx;   /* Hash function context */
-} MD5_context;
+};
 
-extern const MD5_params Curl_DIGEST_MD5[1];
-extern const HMAC_params Curl_HMAC_MD5[1];
+extern const struct MD5_params Curl_DIGEST_MD5[1];
+extern const struct HMAC_params Curl_HMAC_MD5[1];
```
Comment (danielgustafsson) [inline lib/curl_md5.h:50]: Aww, I would've preferred to go the other way and add typedefs.. but that's admittedly mostly Stockholm Syndrome from another codebase I work on. Having a uniform way to reference structs is however a Good Thing, so :+1: on going for it.

Comment (bagder) [reply lib/curl_md5.h:50]: Yeah, I was more going for **the one way** here rather than actually making a stand against typedefs.

**Comment by danielgustafsson** [comment]
> On 5 May 2020, at 10:50, Daniel Stenberg <notifications@github.com> wrote:

> Yeah, I was more going for the one way here rather than actually making a stand against typedefs.

+1

**Comment by bagder** [comment]
Buckle up. I'm going to land this asap. Just pushed a rebase and if things still are fine this will land.

**Comment by bagder** [comment]
Someone emailed me and asked me about the motivation for this work and here's basically my extended motivation.

Using `struct` spelled out in code is more clear than typedef'ed structs in the cases where "struct" is present vs not present. With `struct` present, we know it's a struct. Without it, we don't know what it is (by simply inspecting the name).

But I'm arguing for this change primarily for **consistency**. Our code has a mix of typedef'ed structs and real structs and there's nothing to hint about when we use either approach and I've come to be annoyed by not knowing. It has basically depended on the  mood of the particular author which style that is used in which function.

I want the curl source code to look and feel like it was written by a single author with a single mindset on how code should look like. To make the code clean, easy to read and easy to understand - which then implies consistent. That leads to less bugs and in particular less security problems - a better product. Therefor, I want us to use structs in the code using **one consistent way**.

**Comment by danielgustafsson** [comment]
> On 14 May 2020, at 15:11, Daniel Stenberg <notifications@github.com> wrote:

> ... use structs in the code using one consistent way.

FWIW, this gets a big +1 from me.

**Comment by mkauf** [comment]
Most autobuilds fail, because the ".checksrc" files are not found. The autobuilds perform an "out-of-tree" build.

**Pull request closed without merging**